linked_list/linked_list.py

```diff
@@ -2,78 +2,140 @@
-
+        if self.head == None:
+            return None
+        return self.head.value
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        if self.head == None:
+            self.head = Node(value)
+        else:
+            new_node=Node(value)
+            new_node.next=self.head
+            self.head=new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        node=self.head
+        while node:
+            if self.head.value==value:
+                return True
+            node=self.head.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        counter=0
+        node=self.head
+        while node:
+            counter+=1
+            node=self.head.next
+        return counter
+
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        if self.length()<index:
+            return None
+        counter=0
+        node=self.head
+        while node:
+            if counter==index:
+                return self.head.value
+            node=self.head.next
+            counter+=1
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.length()==0:
+            return None
+        counter=0
+        node=self.head
+        while node:
+            counter+=1
+            if self.head.next == None:
+                tail=Node(self.head)
+                return tail.value
+        return None
+        
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if self.length()==0:
+            self.add_first()
+        else:
+            counter=0
+            node=self.head
+            while node:
+                counter+=1

```

Comment:
👀 Again, how is `counter` helping you make progress towards making `while node` falsy?
```suggestion
            while node:
                counter+=1

```